src/tir/transforms/lower_sparse_tir.cc

```diff
@@ -281,43 +283,120 @@ class IndexTransformer : public StmtExprMutator {
+          block_iters.push_back(SpIterVarToIterVar(sp_it_var, var_map));
+          iter_bindings.push_back(all_loop_vars[i]);
+          has_reduction_var |= sp_it_var->is_reduction;
+        }
+      }
+
+      /* Tag axes in current block as "in-stack". */
+      for (const String&& axis_name : axis_names) {
+        in_stack.insert(std::move(axis_name));
+      }
+
+      /* Update stack. */
+      if (!block_iters.empty()) {
+        block_iters_st.push(std::move(block_iters));
+        iter_bindings_st.push(std::move(iter_bindings));
+        loop_vars_st.push(std::move(loop_vars));
+        if (init_set) {
+          place_init_st.push(false);
+        } else {
+          place_init_st.push(has_reduction_var);
+          init_set |= has_reduction_var;
+        }
+      } else {
+        break;
+      }
+    } while (true);
 
     // Step 4. Generate the read-region and write-retion of the block.
     Array<BufferRegion> reads{nullptr};
     Array<BufferRegion> writes{nullptr};
     GenerateReadWriteRegions(sp_block, &reads, &writes);
 
-    // Step 5. Create the block and block-realize
-    Map<String, ObjectRef> mapping;
-    mapping.Set("sparse", Bool(true));
-    Block block(block_iters, std::move(reads), std::move(writes), sp_block->name, std::move(body),
-                std::move(init), {}, {}, std::move(mapping));
-    BlockRealize block_realize(std::move(iter_bindings), const_true(), std::move(block));
-
-    // Step 6. Create outer loops and the block binding.
-    Stmt loop = GenerateLoops(std::move(block_realize), block_iters, loop_vars);
+    // Step 5. Generate nested blocks and loops from innermost to outermost.
+    String blk_name_hint = sp_block->name;
+    while (!block_iters_st.empty()) {
+      Array<IterVar> block_iters = std::move(block_iters_st.top());
+      Array<PrimExpr> iter_bindings = std::move(iter_bindings_st.top());
+      Array<Var> loop_vars = std::move(loop_vars_st.top());
+      bool place_init = place_init_st.top();
+      block_iters_st.pop();
+      iter_bindings_st.pop();
+      loop_vars_st.pop();
+      place_init_st.pop();
+
+      Map<String, ObjectRef> mapping;
+      mapping.Set("sparse", Bool(true));
+      Block block(/*iter_vars=*/block_iters,
+                  /*reads=*/reads,
+                  /*writes=*/writes,
+                  /*name_hint=*/blk_name_hint,
+                  /*body=*/std::move(body),
+                  /*init=*/place_init ? std::move(init) : NullOpt,
+                  /*alloc_buffers=*/{},
+                  /*match_buffers=*/{},
+                  /*annotations=*/std::move(mapping),
+                  /*span=*/sp_block->span);
+      BlockRealize block_realize(std::move(iter_bindings), const_true(), std::move(block));
+      // Genereate outer loop and the block binding.
+      Stmt loop = GenerateLoops(std::move(block_realize), block_iters, loop_vars);
+      body = loop;
+      blk_name_hint = blk_name_hint + "_outer";
```

Comment:
If there will be multiple blocks, let's use numbering instead of a fixed suffix.